### PR TITLE
PXC-4252: Script that auto-generates the option variable file for WSREP variables

### PR DESCRIPTION
**Pull request by kaushikpuneet07**

### CODE REVIEW

File: pstress/generate_wsrep_options.sh [hunk 1]
```diff
@@ -0,0 +1,239 @@
+#####################################################################################
```
Comment (mohitj1988) [inline pstress/generate_wsrep_options.sh:1]: Fix indentation

**Comment by mohitj1988** [comment]
We can avoid starting the server to fetch PXC and wsrep variables by simply doing
./mysqld --help --verbose

File: pstress/generate_wsrep_options.sh (hunk 1 above)
Comment (kaushikpuneet07) [reply pstress/generate_wsrep_options.sh]: Fixed

**Comment by mohitj1988** [comment]
Paste script ouput for both tarball and manual build.

**Comment by kaushikpuneet07** [comment]
puneet.kaushik@qaserver03:~/pstress/pstress$ ./generate_wsrep_options.sh ~/pxc-8.0/bld/install/
[10:27:34] Working on option '--pxc-encrypt-cluster-traffic' with default value 'TRUE'...
[10:27:34]  > Adding possible values 0 and 1 for option '--pxc-encrypt-cluster-traffic' to the final list...
[10:27:34] Working on option '--pxc-maint-mode' with default value 'DISABLED'...
[10:27:34]  > Adding possible values disabled, shutdown and maintainance for option '--pxc-maint-mode' to the final list...
[10:27:34] Working on option '--pxc-maint-transition-period' with default value '10'...
[10:27:34]  > Adding possible values 0,2,10,16 for option '--pxc-maint-transition-period' to the final list...
[10:27:34] Working on option '--pxc-strict-mode' with default value 'ENFORCING'...
[10:27:34]  > Adding possible values DISABLED, PERMISSIVE, ENFORCING, MASTER for option '--pxc-strict-mode' to the final list...
[10:27:34] Working on option '--wsrep-OSU-method' with default value 'TOI'...
[10:27:34]  > Adding possible values TOI, RSU, NBO for option '--wsrep-OSU-method' to the final list...
[10:27:34] Working on option '--wsrep-RSU-commit-timeout' with default value '5000'...
[10:27:34]  > Adding possible values 5000, 10000, 250000, 31536000000000 for option '--wsrep-RSU-commit-timeout' to the final list...
[10:27:34] Working on option '--wsrep-SR-store' with default value 'table'...
[10:27:34] Option '--wsrep-SR-store' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-applier-FK-checks' with default value 'TRUE'...
[10:27:34] Option '--wsrep-applier-FK-checks' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-applier-UK-checks' with default value 'FALSE'...
[10:27:34] Option '--wsrep-applier-UK-checks' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-applier-threads' with default value '1'...
[10:27:34]  > Adding possible values 1,2,12,24 for option '--wsrep-applier-threads' to the final list...
[10:27:34] Working on option '--wsrep-auto-increment-control' with default value 'TRUE'...
[10:27:34]  > Adding possible values 0,1 for option '--wsrep-auto-increment-control' to the final list...
[10:27:34] Working on option '--wsrep-causal-reads' with default value 'FALSE'...
[10:27:34] Option '--wsrep-causal-reads' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-certification-rules' with default value 'strict'...
[10:27:34]  > Adding possible values STRICT, OPTIMIZED for option '--wsrep-certification-rules' to the final list...
[10:27:34] Working on option '--wsrep-certify-nonPK' with default value 'TRUE'...
[10:27:34]  > Adding possible values 0,1 for option '--wsrep-certify-nonPK' to the final list...
[10:27:34] Working on option '--wsrep-cluster-address' with default value ''...
[10:27:34] Option '--wsrep-cluster-address' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-cluster-name' with default value 'my_wsrep_cluster'...
[10:27:34] Option '--wsrep-cluster-name' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-data-home-dir' with default value '/home/puneet.kaushik/pxc-8.0/bld/install/data/'...
[10:27:34] Option '--wsrep-data-home-dir' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-dbug-option' with default value ''...
[10:27:34] Option '--wsrep-dbug-option' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-debug' with default value 'NONE'...
[10:27:34]  > Adding possible values NONE, SERVER, TRANSACTION, STREAMING , CLIENT  for option '--wsrep-debug' to the final list...
[10:27:34] Working on option '--wsrep-desync' with default value 'FALSE'...
[10:27:34]  > Adding possible values 0,1 for option '--wsrep-desync' to the final list...
[10:27:34] Working on option '--wsrep-dirty-reads' with default value 'FALSE'...
[10:27:34]  > Adding possible values 0,1 for option '--wsrep-dirty-reads' to the final list...
[10:27:34] Working on option '--wsrep-disk-pages-encrypt' with default value 'NONE'...
[10:27:34]  > Adding possible values 0, 1 for option '--wsrep-disk-pages-encrypt' to the final list...
[10:27:34] Working on option '--wsrep-gcache-encrypt' with default value 'NONE'...
[10:27:34]  > Adding possible values 0, 1 for option '--wsrep-gcache-encrypt' to the final list...
[10:27:34] Working on option '--wsrep-ignore-apply-errors' with default value '0'...
[10:27:34] Option '--wsrep-ignore-apply-errors' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-load-data-splitting' with default value 'FALSE'...
[10:27:34]  > Adding possible values 0,1 for option '--wsrep-load-data-splitting' to the final list...
[10:27:34] Working on option '--wsrep-log-conflicts' with default value 'FALSE'...
[10:27:34]  > Adding possible values 0,1 for option '--wsrep-log-conflicts' to the final list...
[10:27:34] Working on option '--wsrep-max-ws-rows' with default value '0'...
[10:27:34]  > Adding possible values 0,1,2,12,254,1023,2047,1048576 for option '--wsrep-max-ws-rows' to the final list...
[10:27:34] Working on option '--wsrep-max-ws-size' with default value '2147483647'...
[10:27:34]  > Adding possible values 0,1,2,12,254,1023,2047,1048576 for option '--wsrep-max-ws-size' to the final list...
[10:27:34] Working on option '--wsrep-min-log-verbosity' with default value '3'...
[10:27:34] Option '--wsrep-min-log-verbosity' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-mode' with default value ''...
[10:27:34]  > Adding possible values Empty, IGNORE_NATIVE_REPLICATION_FILTER_RULES for option '--wsrep-mode' to the final list...
[10:27:34] Working on option '--wsrep-node-address' with default value ''...
[10:27:34] Option '--wsrep-node-address' is logically excluded from being handled by this script...
[10:27:34] Working on option '--wsrep-node-incoming-address' with default value 'AUTO'...
[10:27:35] Option '--wsrep-node-incoming-address' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-node-name' with default value 'qaserver03'...
[10:27:35] Option '--wsrep-node-name' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-notify-cmd' with default value ''...
[10:27:35] Option '--wsrep-notify-cmd' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-on' with default value 'TRUE'...
[10:27:35]  > Adding possible values 0,1 for option '--wsrep-on' to the final list...
[10:27:35] Working on option '--wsrep-provider' with default value 'none'...
[10:27:35] Option '--wsrep-provider' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-provider-options' with default value ''...
[10:27:35] Option '--wsrep-provider-options' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-recover' with default value 'FALSE'...
[10:27:35]  > Adding possible values 0,1 for option '--wsrep-recover' to the final list...
[10:27:35] Working on option '--wsrep-reject-queries' with default value 'NONE'...
[10:27:35]  > Adding possible values NONE, ALL , ALL_KILL for option '--wsrep-reject-queries' to the final list...
[10:27:35] Working on option '--wsrep-replicate-myisam' with default value 'FALSE'...
[10:27:35]  > Adding possible values 0, 1 for option '--wsrep-replicate-myisam' to the final list...
[10:27:35] Working on option '--wsrep-restart-replica' with default value 'FALSE'...
[10:27:35]  > Adding possible values 0, 1 for option '--wsrep-restart-replica' to the final list...
[10:27:35] Working on option '--wsrep-restart-slave' with default value 'FALSE'...
[10:27:35] Option '--wsrep-restart-slave' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-retry-autocommit' with default value '1'...
[10:27:35]  > Adding possible values 0, 1 for option '--wsrep-retry-autocommit' to the final list...
[10:27:35] Working on option '--wsrep-slave-FK-checks' with default value 'TRUE'...
[10:27:35] Option '--wsrep-slave-FK-checks' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-slave-UK-checks' with default value 'FALSE'...
[10:27:35] Option '--wsrep-slave-UK-checks' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-slave-threads' with default value '1'...
[10:27:35] Option '--wsrep-slave-threads' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-sst-allowed-methods' with default value 'xtrabackup-v2'...
[10:27:35] Option '--wsrep-sst-allowed-methods' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-sst-donor' with default value ''...
[10:27:35] Option '--wsrep-sst-donor' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-sst-donor-rejects-queries' with default value 'FALSE'...
[10:27:35]  > Adding possible values 0, 1 for option '--wsrep-sst-donor-rejects-queries' to the final list...
[10:27:35] Working on option '--wsrep-sst-method' with default value 'xtrabackup-v2'...
[10:27:35] Option '--wsrep-sst-method' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-sst-receive-address' with default value 'AUTO'...
[10:27:35] Option '--wsrep-sst-receive-address' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-start-position' with default value '00000000-0000-0000-0000-000000000000:-1'...
[10:27:35] Option '--wsrep-start-position' is logically excluded from being handled by this script...
[10:27:35] Working on option '--wsrep-sync-wait' with default value '0'...
[10:27:35]  > Adding possible values 0,1,2,3,4,5,6,7 for option '--wsrep-sync-wait' to the final list...
[10:27:35] Working on option '--wsrep-trx-fragment-size' with default value '0'...
[10:27:35]  > Adding possible values 0,1,2,12,254,1023,2047,1048576 for option '--wsrep-trx-fragment-size' to the final list...
[10:27:35] Working on option '--wsrep-trx-fragment-unit' with default value 'bytes'...
[10:27:35]  > Adding possible values bytes, rows , statements for option '--wsrep-trx-fragment-unit' to the final list...
Done! Output file: /tmp/mysqld_options_pxc_wsrep_80.txt

**Comment by kaushikpuneet07** [comment]
`puneet.kaushik@qaserver03:~/pstress/pstress$ ./generate_wsrep_options.sh ~/Percona-XtraDB-Cluster_8.0.32-24.2_Linux.x86_64.glibc2.17-minimal
[10:28:04] Working on option '--pxc-encrypt-cluster-traffic' with default value 'TRUE'...
[10:28:04]  > Adding possible values 0 and 1 for option '--pxc-encrypt-cluster-traffic' to the final list...
[10:28:04] Working on option '--pxc-maint-mode' with default value 'DISABLED'...
[10:28:04]  > Adding possible values disabled, shutdown and maintainance for option '--pxc-maint-mode' to the final list...
[10:28:04] Working on option '--pxc-maint-transition-period' with default value '10'...
[10:28:04]  > Adding possible values 0,2,10,16 for option '--pxc-maint-transition-period' to the final list...
[10:28:04] Working on option '--pxc-strict-mode' with default value 'ENFORCING'...
[10:28:04]  > Adding possible values DISABLED, PERMISSIVE, ENFORCING, MASTER for option '--pxc-strict-mode' to the final list...
[10:28:04] Working on option '--wsrep-OSU-method' with default value 'TOI'...
[10:28:04]  > Adding possible values TOI, RSU, NBO for option '--wsrep-OSU-method' to the final list...
[10:28:04] Working on option '--wsrep-RSU-commit-timeout' with default value '5000'...
[10:28:04]  > Adding possible values 5000, 10000, 250000, 31536000000000 for option '--wsrep-RSU-commit-timeout' to the final list...
[10:28:04] Working on option '--wsrep-SR-store' with default value 'table'...
[10:28:04] Option '--wsrep-SR-store' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-applier-FK-checks' with default value 'TRUE'...
[10:28:04] Option '--wsrep-applier-FK-checks' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-applier-UK-checks' with default value 'FALSE'...
[10:28:04] Option '--wsrep-applier-UK-checks' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-applier-threads' with default value '1'...
[10:28:04]  > Adding possible values 1,2,12,24 for option '--wsrep-applier-threads' to the final list...
[10:28:04] Working on option '--wsrep-auto-increment-control' with default value 'TRUE'...
[10:28:04]  > Adding possible values 0,1 for option '--wsrep-auto-increment-control' to the final list...
[10:28:04] Working on option '--wsrep-causal-reads' with default value 'FALSE'...
[10:28:04] Option '--wsrep-causal-reads' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-certification-rules' with default value 'strict'...
[10:28:04]  > Adding possible values STRICT, OPTIMIZED for option '--wsrep-certification-rules' to the final list...
[10:28:04] Working on option '--wsrep-certify-nonPK' with default value 'TRUE'...
[10:28:04]  > Adding possible values 0,1 for option '--wsrep-certify-nonPK' to the final list...
[10:28:04] Working on option '--wsrep-cluster-address' with default value ''...
[10:28:04] Option '--wsrep-cluster-address' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-cluster-name' with default value 'my_wsrep_cluster'...
[10:28:04] Option '--wsrep-cluster-name' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-data-home-dir' with default value '/home/puneet.kaushik/Percona-XtraDB-Cluster_8.0.32-24.2_Linux.x86_64.glibc2.17-minimal/data/'...
[10:28:04] Option '--wsrep-data-home-dir' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-dbug-option' with default value ''...
[10:28:04] Option '--wsrep-dbug-option' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-debug' with default value 'NONE'...
[10:28:04]  > Adding possible values NONE, SERVER, TRANSACTION, STREAMING , CLIENT  for option '--wsrep-debug' to the final list...
[10:28:04] Working on option '--wsrep-desync' with default value 'FALSE'...
[10:28:04]  > Adding possible values 0,1 for option '--wsrep-desync' to the final list...
[10:28:04] Working on option '--wsrep-dirty-reads' with default value 'FALSE'...
[10:28:04]  > Adding possible values 0,1 for option '--wsrep-dirty-reads' to the final list...
[10:28:04] Working on option '--wsrep-disk-pages-encrypt' with default value 'NONE'...
[10:28:04]  > Adding possible values 0, 1 for option '--wsrep-disk-pages-encrypt' to the final list...
[10:28:04] Working on option '--wsrep-gcache-encrypt' with default value 'NONE'...
[10:28:04]  > Adding possible values 0, 1 for option '--wsrep-gcache-encrypt' to the final list...
[10:28:04] Working on option '--wsrep-ignore-apply-errors' with default value '0'...
[10:28:04] Option '--wsrep-ignore-apply-errors' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-load-data-splitting' with default value 'FALSE'...
[10:28:04]  > Adding possible values 0,1 for option '--wsrep-load-data-splitting' to the final list...
[10:28:04] Working on option '--wsrep-log-conflicts' with default value 'FALSE'...
[10:28:04]  > Adding possible values 0,1 for option '--wsrep-log-conflicts' to the final list...
[10:28:04] Working on option '--wsrep-max-ws-rows' with default value '0'...
[10:28:04]  > Adding possible values 0,1,2,12,254,1023,2047,1048576 for option '--wsrep-max-ws-rows' to the final list...
[10:28:04] Working on option '--wsrep-max-ws-size' with default value '2147483647'...
[10:28:04]  > Adding possible values 0,1,2,12,254,1023,2047,1048576 for option '--wsrep-max-ws-size' to the final list...
[10:28:04] Working on option '--wsrep-min-log-verbosity' with default value '3'...
[10:28:04] Option '--wsrep-min-log-verbosity' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-mode' with default value ''...
[10:28:04]  > Adding possible values Empty, IGNORE_NATIVE_REPLICATION_FILTER_RULES for option '--wsrep-mode' to the final list...
[10:28:04] Working on option '--wsrep-node-address' with default value ''...
[10:28:04] Option '--wsrep-node-address' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-node-incoming-address' with default value 'AUTO'...
[10:28:04] Option '--wsrep-node-incoming-address' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-node-name' with default value 'qaserver03'...
[10:28:04] Option '--wsrep-node-name' is logically excluded from being handled by this script...
[10:28:04] Working on option '--wsrep-notify-cmd' with default value ''...
[10:28:05] Option '--wsrep-notify-cmd' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-on' with default value 'TRUE'...
[10:28:05]  > Adding possible values 0,1 for option '--wsrep-on' to the final list...
[10:28:05] Working on option '--wsrep-provider' with default value 'none'...
[10:28:05] Option '--wsrep-provider' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-provider-options' with default value ''...
[10:28:05] Option '--wsrep-provider-options' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-recover' with default value 'FALSE'...
[10:28:05]  > Adding possible values 0,1 for option '--wsrep-recover' to the final list...
[10:28:05] Working on option '--wsrep-reject-queries' with default value 'NONE'...
[10:28:05]  > Adding possible values NONE, ALL , ALL_KILL for option '--wsrep-reject-queries' to the final list...
[10:28:05] Working on option '--wsrep-replicate-myisam' with default value 'FALSE'...
[10:28:05]  > Adding possible values 0, 1 for option '--wsrep-replicate-myisam' to the final list...
[10:28:05] Working on option '--wsrep-restart-replica' with default value 'FALSE'...
[10:28:05]  > Adding possible values 0, 1 for option '--wsrep-restart-replica' to the final list...
[10:28:05] Working on option '--wsrep-restart-slave' with default value 'FALSE'...
[10:28:05] Option '--wsrep-restart-slave' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-retry-autocommit' with default value '1'...
[10:28:05]  > Adding possible values 0, 1 for option '--wsrep-retry-autocommit' to the final list...
[10:28:05] Working on option '--wsrep-slave-FK-checks' with default value 'TRUE'...
[10:28:05] Option '--wsrep-slave-FK-checks' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-slave-UK-checks' with default value 'FALSE'...
[10:28:05] Option '--wsrep-slave-UK-checks' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-slave-threads' with default value '1'...
[10:28:05] Option '--wsrep-slave-threads' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-sst-allowed-methods' with default value 'xtrabackup-v2'...
[10:28:05] Option '--wsrep-sst-allowed-methods' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-sst-donor' with default value ''...
[10:28:05] Option '--wsrep-sst-donor' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-sst-donor-rejects-queries' with default value 'FALSE'...
[10:28:05]  > Adding possible values 0, 1 for option '--wsrep-sst-donor-rejects-queries' to the final list...
[10:28:05] Working on option '--wsrep-sst-method' with default value 'xtrabackup-v2'...
[10:28:05] Option '--wsrep-sst-method' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-sst-receive-address' with default value 'AUTO'...
[10:28:05] Option '--wsrep-sst-receive-address' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-start-position' with default value '00000000-0000-0000-0000-000000000000:-1'...
[10:28:05] Option '--wsrep-start-position' is logically excluded from being handled by this script...
[10:28:05] Working on option '--wsrep-sync-wait' with default value '0'...
[10:28:05]  > Adding possible values 0,1,2,3,4,5,6,7 for option '--wsrep-sync-wait' to the final list...
[10:28:05] Working on option '--wsrep-trx-fragment-size' with default value '0'...
[10:28:05]  > Adding possible values 0,1,2,12,254,1023,2047,1048576 for option '--wsrep-trx-fragment-size' to the final list...
[10:28:05] Working on option '--wsrep-trx-fragment-unit' with default value 'bytes'...
[10:28:05]  > Adding possible values bytes, rows , statements for option '--wsrep-trx-fragment-unit' to the final list...
Done! Output file: /tmp/mysqld_options_pxc_wsrep_80.txt
`

File: pstress/generate_wsrep_options.sh [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  echo "[$(date +'%T')] $1" >> /tmp/generate_wsrep_options.log
+}
+
+
+# Extract all wsrep options, their default values into temp file
+$BASEDIR/bin/mysqld --no-defaults --help --verbose | grep "^pxc-*" 2>/dev/null > ${TEMP_FILE}
+$BASEDIR/bin/mysqld --no-defaults --help --verbose | grep "^wsrep-*" 2>/dev/null >> ${TEMP_FILE}
+
+# List of wsrep variables which must not be changed.
+EXCLUDED_LIST=( --wsrep-applier-FK-checks --wsrep-applier-UK-checks --wsrep-causal-reads --wsrep-cluster-address --wsrep-cluster-name --wsrep-data-home-dir --wsrep-dbug-option --wsrep-drupal-282555-workaround --wsrep-forced-binlog-format --wsrep-ignore-apply-errors --wsrep-min-log-verbosity --wsrep-node-address --wsrep-node-incoming-address --wsrep-node-name --wsrep-notify-cmd --wsrep-provider --wsrep-provider-options --wsrep-restart-slave --wsrep-slave-FK-checks --wsrep-slave-threads --wsrep-slave-UK-checks --wsrep-sst-donor --wsrep-sst-method --wsrep-sst-receive-address --wsrep-start-position --wsrep-disk-pages-encrypt --wsrep-gcache-encrypt --wsrep-SR-store --wsrep-sst-allowed-methods)
```
[the remark this answers is not in the document]
Comment (kaushikpuneet07) [reply pstress/generate_wsrep_options.sh:38]: Done

**Review by mohitj1988 — APPROVED** [review]
Good to be merged

**Pull request merged**